runtime/direct_process.go

```diff
@@ -0,0 +1,283 @@
+package runtime
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+	"os"
+	"path"
+	"path/filepath"
+	"sync"
+	"syscall"
+
+	"github.com/docker/containerd/specs"
+	"github.com/docker/containerd/subreaper"
+	"github.com/docker/containerd/subreaper/exec"
+	"github.com/docker/docker/pkg/term"
+	"github.com/opencontainers/runc/libcontainer"
+)
+
+type directProcess struct {
+	*process
+	sync.WaitGroup
+
+	io          stdio
+	console     libcontainer.Console
+	consolePath string
+	exec        bool
+	checkpoint  string
+	specs       *specs.Spec
+}
+
+func newDirectProcess(config *processConfig) (*directProcess, error) {
+	lp, err := newProcess(config)
+	if err != nil {
+		return nil, err
+	}
+
+	return &directProcess{
+		specs:      config.spec,
+		process:    lp,
+		exec:       config.exec,
+		checkpoint: config.checkpoint,
+	}, nil
+}
+
+func (d *directProcess) CloseStdin() error {
+	if d.io.stdin != nil {
+		return d.io.stdin.Close()
+	}
+	return nil
+}
+
+func (d *directProcess) Resize(w, h int) error {
+	if d.console == nil {
+		return nil
+	}
+	ws := term.Winsize{
+		Width:  uint16(w),
+		Height: uint16(h),
+	}
+	return term.SetWinsize(d.console.Fd(), &ws)
+}
+
+func (d *directProcess) openIO() (*os.File, *os.File, *os.File, error) {
+	uid, gid, err := getRootIDs(d.specs)
+	if err != nil {
+		return nil, nil, nil, err
+	}
+
+	if d.spec.Terminal {
+		console, err := libcontainer.NewConsole(uid, gid)
+		if err != nil {
+			return nil, nil, nil, err
+		}
+		d.console = console
+		d.consolePath = console.Path()
+		stdin, err := os.OpenFile(d.stdio.Stdin, syscall.O_RDONLY, 0)
+		if err != nil {
+			return nil, nil, nil, err
+		}
+		go io.Copy(console, stdin)
+		stdout, err := os.OpenFile(d.stdio.Stdout, syscall.O_RDWR, 0)
+		if err != nil {
+			return nil, nil, nil, err
+		}
+		d.Add(1)
+		go func() {
+			io.Copy(stdout, console)
```

Comment:
There is currently a bug in which this line will never exit in certain situations even after the console is closed in delete().